src/common/Finisher.h

```diff
@@ -139,34 +117,11 @@ class Finisher {
 
   /// Construct an anonymous Finisher.
   /// Anonymous finishers do not log their queue length.
-  explicit Finisher(CephContext *cct_) :
-    cct(cct_), finisher_lock(ceph::make_mutex("Finisher::finisher_lock")),
-    finisher_stop(false), finisher_running(false), finisher_empty_wait(false),
-    thread_name("fn_anonymous"), logger(0),
-    finisher_thread(this) {}
+  explicit Finisher(CephContext *cct_);
 
   /// Construct a named Finisher that logs its queue length.
-  Finisher(CephContext *cct_, std::string name, std::string tn) :
-    cct(cct_), finisher_lock(ceph::make_mutex("Finisher::" + name)),
-    finisher_stop(false), finisher_running(false), finisher_empty_wait(false),
-    thread_name(tn), logger(0),
-    finisher_thread(this) {
-    PerfCountersBuilder b(cct, std::string("finisher-") + name,
-			  l_finisher_first, l_finisher_last);
-    b.add_u64(l_finisher_queue_len, "queue_len");
-    b.add_time_avg(l_finisher_complete_lat, "complete_latency");
-    logger = b.create_perf_counters();
-    cct->get_perfcounters_collection()->add(logger);
-    logger->set(l_finisher_queue_len, 0);
-    logger->set(l_finisher_complete_lat, 0);
-  }
-
-  ~Finisher() {
-    if (logger && cct) {
-      cct->get_perfcounters_collection()->remove(logger);
-      delete logger;
-    }
-  }
+  Finisher(CephContext *cct_, std::string_view name, std::string_view tn);
```

Comment:
I'm surprise
```suggestion
  Finisher(CephContext *cct_, std::string_view name, std::string&& tn);
```

+ move?

Reply:
> All callers I saw pass a C string, and since you prefer string_view over C strings,

As always, depends on context.

>  I did exactly that; but why let the compiler generate code in all callers to construct a fat std::string?

Because we plan to use the string to construct `Finisher::thread_name`. So providing the option to move a `std::string&&` is ideal.

(Sorry my original comment was truncated somehow while writing it.)

Reply:
> So providing the option to move a `std::string&&` is ideal.

That is only a benefit if somebody actually passes a `std::string`. As long as all callers pass a C string, the C string is the most efficient way to do it (and `std::string_view` is the second most efficient way), because coercing everything into a `std::string` means the compiler duplicates some machine code in the binary, one copy of it for every call site.

But I changed it to `std::string&&`.